### DISCUSSION TITLE
Quamina Priorities

### DISCUSSION THREAD

**Discussion by timbray** [timbray/quamina#321] [General]
For a long time I've had the states-explosion problem at the front of my mind and now that it's solved, I can start to think of other things to do.  Some of the most obvious forward directions are:

# More optimization

Q's performance is already good, except for the corner case with wildcards where, if you combine a few thousand, you can slow Q down by a couple orders of magnitude.  Fixing this is probably the top optimization prioritization item.  There's no guarantee of a fix (AWS-Event-Ruler still suffers from a flavor of this problem) but I have some strong ideas, one of them being turning the NFA into a DFA.

# Implement Ruler-compatible `wildcard` patterns

These are the same as our `shellstyle` patterns but have escaping, so you can say:
```json
{"x": [{"wildcard":"Hey \\*Joe\\*, get the *"}]}`
```
So, the task is to generalize \\-escaping.

# Implement a bunch of Ruler-compatibility changes 

There are a bunch of Issues open for this.  These should be fairly straightforward with the newer, cleaner, NFA-building approach.  If someone feels like they'd like to do feature-building, this is a strong candidate.  

One that's special is the general-purpose numeric value handling. There's code sketched in but never actually turned on.  

# Implement general regular expressions

This will be more intellectually challenging but probably intense fun.  It also has the advantage that if you do this, probably many of the Ruler-compatibility things can be expressed as a regular expression.  

# What do you think?

I have opinions but would like to hear others.

**Comment by embano1** [comment]
Regarding 

> These are the same as our shellstyle patterns but have escaping

Basically you're saying be 100% `aws/ruler` compatible to allow:

> Wildcard matches only work on string-valued fields. A single value can contain zero to many wildcard characters, but consecutive wildcard characters are not allowed. To match the asterisk character specifically, a wildcard character can be escaped with a backslash. Two consecutive backslashes (i.e. a backslash escaped with a backslash) represents the actual backslash character. A backslash escaping any character other than asterisk or backslash is not allowed.

**Reply by timbray** [reply]
Exactly.  The idea is, once Q has compatible `wildcard` matching we'll deprecate (but never remove) `shellstyle`.

For the same reason, I have another PR coming to do the first steps of numeric matching.  Q already had `number.go` (not yet used in practice) which is similar to Ruler's `ComparableNumber` but with 2 digits more precision.  But I'm going to cut it back to have the exact same semantics as Ruler, because I think compatibility is a big deal.

**Reply by embano1** [reply]
Got it!
